### PR TITLE
Add ocr2:inspect:responses command to gauntlet-solana

### PR DESCRIPTION
**Pull request by kylesmartin**

### CODE REVIEW

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 1]
```diff
@@ -1,11 +1,25 @@
...
 import { SolanaCommand, TransactionResponse } from '@chainlink/gauntlet-solana'
 import { PublicKey } from '@solana/web3.js'
 import { CONTRACT_LIST, getContract } from '../../../../lib/contracts'
-import { inspection } from '@chainlink/gauntlet-core/dist/utils'
+import { inspection, logger, BN } from '@chainlink/gauntlet-core/dist/utils'
+import { ORACLES_MAX_LENGTH } from '../../../../lib/constants'
 
 type Input = {}
 
+type NewTransmission = {
```
Comment (krebernisak) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: Let's group types like this not under commands that depend on them, but as general OCR2  types somewhere like `contracts/ocr2/types.ts`.

Long term we'll need to think about the overall structure of our TS toolchain (SDK + Gautlet). One of the suggestions would be to move types like this and core functionality to our TS SDK (`@chainlink/solana-sdk`) and then build Gauntlet on top of it.

cc @RodrigoAD

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 2]
```diff
@@ -25,13 +39,118 @@ export default class OCR2InspectResponses extends SolanaCommand {
...
+      onChainState.config.latestConfigBlockNumber.toNumber(),
+      ocr2.programId,
+      state,
+      ocr2.idl,
+      onChainState.oracles.xs
+        .map((oracle) => oracle.transmitter)
+        .filter((transmitter) => {
+          return transmitter._bn != 0
+        }),
+    )
```
Comment (krebernisak) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]:
```suggestion
    const blockNumber = onChainState.config.latestConfigBlockNumber.toNumber()
    const transmitters = onChainState.oracles.xs
        .map((oracle) => oracle.transmitter)
        .filter((transmitter) => transmitter._bn != 0 ),
    this.getLatestTransmissionEvent(blockNumber, ocr2.programId, state, ocr2.idl, transmitters)
```

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 3]
```diff
@@ -25,13 +39,118 @@ export default class OCR2InspectResponses extends SolanaCommand {
...
+    const eventParser = new EventParser(programId, coder)
+    // Get block
+    let block = await this.provider.connection.getBlock(blockNumber)
+    if (!block) {
+      throw new Error('Block not found. Could not find latest block number in config')
+    }
+    // Iterate over all transactions in block
+    block.transactions.forEach((transaction) => {
+      // Get list of accounts keys associated with txn
+      let accountKeys = transaction.transaction.message.accountKeys.map((key) => `${key}`)
```
Comment (krebernisak) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: All these `let`s should be redeclared as `const`s

**Review by krebernisak — COMMENTED** [review]
Looks pretty good! Please also contribute documentation for the new command. Thx!

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 4]
```diff
@@ -25,13 +39,118 @@ export default class OCR2InspectResponses extends SolanaCommand {
...
+
+    logger.info(`Latest Config: 
+    - Latest Transmitter: ${onChainState.config.latestTransmitter}
+    - Latest Aggregator Round ID: ${onChainState.config.latestAggregatorRoundId}
+    - Latest Config Digest: ${onChainState.config.latestConfigDigest}
+    - Latest Config Block Number: ${onChainState.config.latestConfigBlockNumber}
+    `)
+
+    this.getLatestTransmissionEvent(
+      onChainState.config.latestConfigBlockNumber.toNumber(),
```
Comment (RodrigoAD) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: - We want to inspect the latest transmission, for that we should search for the latest `NewTransmissions` event emitted. Now you are searching within the block where the latest config was set, this has 2 issues:
  - A new transmission was not necessarily included there
  - Most probably won't be the latest transmission in the contract
- `getLatestTransmissionEvent` should just get the transmission event. It should not log it or any other side effect.
- We can use anchor to parse the event, take a look at our SDK for parsing transmissions: https://github.com/smartcontractkit/chainlink-solana/blob/develop/ts/src/index.ts#L42

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 5]
```diff
@@ -25,13 +39,123 @@ export default class OCR2InspectResponses extends SolanaCommand {
...
+  ) => {
+    // Counts number of NewTransmissions found
+    var newTransmissionCount: number = 0
+    // Tracks the last sig checked
+    var lastSigChecked: string | undefined = undefined
+    // Number of transactions to check at once
+    const batchSize = 100
+    // Define coder and event parser
+    const coder = new BorshCoder(idl)
+    const eventParser = new EventParser(programId, coder)
```
Comment (RodrigoAD) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]:
```suggestion
    const eventParser = new EventParser(program.programId, program.coder)
```

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 6]
```diff
@@ -25,13 +39,123 @@ export default class OCR2InspectResponses extends SolanaCommand {
...
+    // Loop until transmission found
+    while (newTransmissionCount == 0) {
+      // Get {batchSize} latest sigs
+      const sigs = await this.provider.connection.getSignaturesForAddress(programAccount, {
+        limit: batchSize,
+        before: lastSigChecked,
+      })
+      // Get the txns associated with the sigs
+      const txns = await this.provider.connection.getParsedTransactions(sigs.map((sig) => sig.signature))
```
Comment (RodrigoAD) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: Let's abstract this into a  `getLatestTxs = (amount: number)`

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 7]
```diff
@@ -25,13 +39,123 @@ export default class OCR2InspectResponses extends SolanaCommand {
...
+      txns.forEach((txn) => {
+        // Check txns with logs
+        if (txn && txn.meta && txn.meta.logMessages) {
+          // Parse logs
+          eventParser.parseLogs(txn.meta.logMessages, (event: Event) => {
```
Comment (RodrigoAD) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: Let me come up with a non callback thing

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 8]
```diff
@@ -15,10 +45,33 @@ export default class OCR2InspectResponses extends SolanaCommand {
...
+      return undefined
+    }
+
+    const rdd = RDD.load(network, rddPath)
+    const aggregator = RDD.loadAggregator(this.args[0], network, rddPath)
+    const aggregatorOracles: Oracle[] = aggregator.oracles.map((o) => {
+      return {
+        transmitter: rdd.operators[o.operator].ocrNodeAddress[0],
+        payee: rdd.operators[o.operator].adminAddress,
+        signer: rdd.operators[o.operator].ocr2OnchainPublicKey[0].substring(14),
```
Comment (RodrigoAD) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: `signer` and `payee` don't seem to be used

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 9]
```diff
@@ -2,9 +2,39 @@ import { Result } from '@chainlink/gauntlet-core'
...
+type Oracle = {
+  transmitter: string
+  website: string
+  name: string
+  apis: string[]
+  observation?: number
+}
+
+// Returns a formatted oracle log given an address, a command input, a label, and a start tab,
+const oracleLog = (address: string, label: string, startTab: number, input?: Input) => {
```
Comment (RodrigoAD) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: We can simplify and make a bit clearer the interface of this. Receiving the oracle to print or the list of oracles seems cleaner. We can give a default value to the `startTab` as well.
```suggestion
const makeOracleLog = (address: string, label: string, oracles: Oracle[], startTab = 2) => {
```

File: gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts [hunk 10]
```diff
@@ -2,9 +2,39 @@ import { Result } from '@chainlink/gauntlet-core'
...
+  description: string
+  aggregatorOracles: Oracle[]
+}
+
+type Oracle = {
+  transmitter: string
+  website: string
+  name: string
+  apis: string[]
+  observation?: number
```
Comment (RodrigoAD) [inline gauntlet/packages/gauntlet-solana-contracts/src/commands/contracts/ocr2/inspection/inspectResponses.ts]: This is not used either
```suggestion
```

**Review by RodrigoAD — COMMENTED** [review]
It looks really good, just some minor changes and we can merge

**Review by RodrigoAD — APPROVED** [review]
I think is good to go! Thanks @kylesmartin

**Pull request merged**